kong/plugins/session/session.lua

```diff
@@ -62,21 +63,30 @@ function _M.retrieve_session_data(s)
     return nil, nil
   end
 
-  return s.data[1], s.data[2]
+  return s.data[1], s.data[2], s.data[3]
 end
 
 
 --- Store the session data for usage in kong plugins
 -- @param s - the session
 -- @param consumer - the consumer id
 -- @param credential - the credential id or potentially just the consumer id
-function _M.store_session_data(s, consumer_id, credential_id)
+-- @param groups - table of authenticated_groups e.g. ["1" = "group1", "group1"]
+function _M.store_session_data(s, consumer_id, credential_id, groups)
   if not s then
     return
   end
 
   s.data[1] = consumer_id
   s.data[2] = credential_id
+
+  if groups then
+    -- ensure serialized table is set to array so data is not stored as json
+    -- with mixed type (e.g. object _and_ array).
+    setmetatable(groups, cjson.array_mt)
```

Comment:
this is tricky, since this is a cached entity? violating the read-only nature of cached entities?

Reply:
nvm, it is a copy. But then, shouldn't the metatable be set when array is created?